src/service/layers-impl.js

```diff
@@ -0,0 +1,869 @@
+     *
+     * @private {boolean}
+     */
+    this.needsRemeasure_ = true;
+
+    /**
+     * The current cached size of the element.
+     *
+     * @private {!SizeDef}
+     */
+    this.size_ = sizeWh(0, 0);
+
+    /**
+     * The current cached **offset** position of the element, relative to the
+     * parent layer. Note that offsets are constant, and unchanged by the
+     * current scroll position of the parent.
+     *
+     * @private {!PositionDef}
+     */
+    this.position_ = positionLt(0, 0);
+
+    /**
+     * Whether the element defines a layer with a new coordinate system.
+     * Elements may freely transition to/from being a layer, causing child
+     * elements to redefine their offset positions.
+     *
+     * @private {boolean}
+     */
+    this.isLayer_ = false;
+
+    /**
+     * Whether the layer is a "root" scrolling layer. Root scrollers have
+     * special properties, mainly that they can never be un-declared (they will
+     * always exist) and that their offset positions are not defined by
+     * scrollTop (DOM APIs are inconsistent between a root scroller and an
+     * overflow scroller).
+     *
+     * @private {boolean}
+     */
+    this.isRootLayer_ = false;
+
+    /**
+     * Whether this layer needs to remeasure its scrollTop/Left position during
+     * the next position calculation. This is done lazily so that scroll
+     * performance is not hampered.
+     *
+     * @private {boolean}
+     */
+    this.needsScrollRemeasure_ = false;
+
+    /** @private {number} */
+    this.scrollLeft_ = 0;
+
+    /** @private {number} */
+    this.scrollTop_ = 0;
+
+    /**
+     * The child LayoutElements of this layer. Only a layer (which means it
+     * defines a coordinate system) may have children, even if the element
+     * _contains_ other elements.
+     *
+     * @const @private {!Array<!LayoutElement>}
+     */
+    this.children_ = [];
+  }
+
+  /**
+   * Gets the element's LayoutElement instance.
+   *
+   * @param {!Element}
+   * @return {!LayoutElement}
+   */
+  static for(element) {
+    return dev().assert(LayoutElement.forOptional(element));
+  }
+
+  /**
+   * Gets the element's LayoutElement instance, if the element has one.
+   *
+   * @param {!Element}
+   * @return {?LayoutElement}
+   */
+  static forOptional(element) {
+    return element[LAYOUT_PROP] || null;
+  }
+
+  /**
+   * Finds the element's parent layer
+   *
+   * If the element is itself a layer, it still looks in the element's ancestry
+   * for a parent layer.
+   *
+   * @param {!Node} node
+   * @param {opt_force=} Whether to force a re-lookup
+   * @return {?LayoutElement}
+   */
+  static getParentLayer(node, opt_force) {
+    if (!opt_force) {
+      const layout = LayoutElement.forOptional(node);
+      if (layout) {
+        return layout.getParentLayer();
+      }
+    }
+
+    const win = node.ownerDocument.defaultView;
+    let op = node;
+    for (let el = node; el; el = el.parentNode) {
```

Comment:
I lost track what we do with FIE and shadow docs here? Were you planning to support them in a follow-up PR? At a minimum, it might be worth checking here for the ampdoc's rootNode, so you don't accidentally go outside the shadow doc.

Reply:
https://github.com/ampproject/amphtml/issues/12554